File: 53-maximum-subarray/53-maximum-subarray.scala
object Solution {
    def maxSubArray(nums: Array[Int]): Int = {
        var res = nums.reduce(_ max _)
        var cur = 0
        
        if (res >= 0) {
            for(n <- nums) {
                cur += n
                cur = cur max 0
                res = res max cur
            }
        }
        
        res
    }
}

File: 53-maximum-subarray/53-maximum-subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int result = nums[0];
        int current = 0;
        
        for(int n : nums) {
            int sum = current + n;
            result = Math.max(result, sum);
            current = Math.max(sum, 0);
        }
        
        return result;
    }
}